esp32-wroom-rp/src/tcp_client.rs

```diff
@@ -32,13 +33,71 @@ impl Format for TcpError {
     }
 }
 
+pub trait Connect<'a, S, B, C> {
+    fn connect<F: Fn(&TcpClient<'a, B, C>), D: DelayMs<u16>>(
+        &mut self,
+        server: S,
+        port: Port,
+        mode: TransportMode,
+        delay: &mut D,
+        f: F,
+    ) -> Result<(), Error>;
+}
+
 pub struct TcpClient<'a, B, C> {
     pub(crate) protocol_handler: &'a mut NinaProtocolHandler<B, C>,
     pub(crate) socket: Option<Socket>,
     pub(crate) server_ip_address: Option<IpAddress>,
     pub(crate) port: Port,
     pub(crate) mode: TransportMode,
-    pub(crate) server_hostname: Option<&'a str>,
+    pub(crate) server_hostname: Option<String<MAX_HOSTNAME_LENGTH>>,
+}
+
+impl<'a, B, C> Connect<'a, IpAddress, B, C> for TcpClient<'a, B, C>
+where
+    B: Transfer<u8>,
+    C: EspControlInterface,
+{
+    fn connect<F: Fn(&TcpClient<'a, B, C>), D: DelayMs<u16>>(
+        &mut self,
+        ip: IpAddress,
+        port: Port,
+        mode: TransportMode,
+        delay: &mut D,
+        f: F,
+    ) -> Result<(), Error> {
+        let socket = self.get_socket().unwrap_or_default();
```

Comment:
I'm curious what you think about using the `?` operator here? 
My thought process being that if this winds up returning a default (I'm assuming it's `0u8`) then the rest of the code will think that `get_socket` succeeded since the TCP socket that we receive from NINA firmware is always `0u8` anyway. Things will fail downstream anyway in that scenario but it might be better to fail earlier.

Reply:
That's a really good thought and I think that makes sense indeed for exactly the point you made.